pymeasure/instruments/hp/hp4145x.py

```diff
@@ -0,0 +1,1116 @@
+            Only works with mode == 'USER_MODE' of instrument. In 'SYSTEM_MODE' no
+            direct access is allowed.
+        """
+        if self.parent.mode == 'USER_MODE':
+            value = float(self.ask(f"TV {self.id + 4}"))
+        else:
+            raise RuntimeError("In SYSTEM_MODE not allowed")
+
+        self.check_get_errors()
+        return value
+
+    @voltage.setter
+    def voltage(self, value):
+        raise LookupError("Property can not be set.")
+
+    @property
+    def disabled(self):
+        """
+        Set the disabling of the respective channel.
+        """
+        raise LookupError("Property can not be read.")
+
+    @disabled.setter
+    def disabled(self, value):
+        self._disabled = True
+        if not value:
+            raise ValueError(
+                "Channel can only be disabled, it gets enabled implicitly with the other functions")
+
+        self.write("DE VM {ch}")
+
+    @property
+    def voltage_name(self):
+        """
+        Control the voltage name for `SYSTEM_MODE`'s channel definition.
+
+        Defaults to VM[1..2]. Must be unique otherwise the instrument returns a 'PROGRAM_ERROR'.
+
+        Returns not the actual state of the instrument but the state of the object.
+        """
+        raise LookupError("Property can not be read.")
+
+    @voltage_name.setter
+    def voltage_name(self, value):
+        if len(value.strip()) > 6 or len(value.strip()) < 1:
+            raise ValueError("Name must be less than six and at least one characters")
+
+        self._voltage_name = value.upper()
+        self._disabled = False
+
+        if not self.manual_flush:
+            self.write("DE VM %d, '%s'" % (self.id, self._voltage_name))
+            self.check_errors()
+
+    def flush_channel_definition(self):
+        """
+        Flushes the channel definition manually in case the :attr:`manual_flush` option is set.
+        """
+        if not self._disabled:
+            self.write("DE VM %d, '%s'" % (self.id, self._voltage_name))
+            self.check_errors()
+
+
+class SMU(Channel):
+    _voltage_range = 0
+    _voltage_ranges = {0: 0, 20: 1, 40: 2, 100: 3}
+    _current_range = 0
+    _current_ranges = {0: 0, 1e-9: 1, 10e-9: 2, 100e-9: 3,
+                       1e-6: 4, 10e-6: 5, 100e-6: 6, 1e-3: 7, 10e-3: 8, 100e-3: 9}
+
+    _current_max = 100e-3
+    _current_compliance = 0.0
+
+    _voltage_max = 100
+    _voltage_compliance = 0.0
+
+    _voltage_name = ''
+    _current_name = ''
+
+    _channel_function = 3
+    _channel_functions = {"VAR1": 1, "VAR2": 2, "CONST": 3, "VAR1'": 4}
+
+    _channel_mode = 3
+    _channel_modes = {'V': 1, 'I': 2, 'COM': 3}
+
+    _voltage = 0
+    _current = 0
+
+    _disabled = False
+
+    manual_flush = False
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._voltage_name = f"V{self.id}"
+        self._current_name = f"I{self.id}"
+
+    @property
+    def voltage(self):
```

Comment:
Is it the actual voltage or the voltage setpoint? In the latter case, `voltage_setpoint` might be a better name.
Most of the time we use `setpoint` to distinguish the setpoint from the actual measurement.

Reply:
You say, that you measure the voltage (i.e. read only), but you also have a setter, so do you measure it or do you control it?
If you control it, it is a setpoint, isn't it?

Reply:
With use model you mean, that you have `voltage_setpoint = 5` and then you read the actual output voltage with `voltage == 5`?

We chose the setpoint version as for some instruments, `voltage` controls the setpoint (get and set), and sometimes it measures the actual voltage. For consistency and clarity in this library, we should use it here as well.